cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlStructuredTableRegistry.java

```diff
@@ -16,40 +16,181 @@
-    return specMap.get(tableId);
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable registry = context.getTable(REGISTRY);
+      Optional<StructuredRow> optional =
+        registry.read(Collections.singleton(Fields.stringField(TABLE_NAME_FIELD, tableId.getName())));
+      if (!optional.isPresent()) {
+        return null;
+      }
+      String specString = optional.get().getString(TABLE_SPEC_FIELD);
+      LOG.trace("Got specification {} from registry", specString);
+      return GSON.fromJson(specString, StructuredTableSpecification.class);
+    });
   }
 
   public void removeSpecification(StructuredTableId tableId) {
-    specMap.remove(tableId);
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable registry = context.getTable(REGISTRY);
+      registry.delete(Collections.singleton(Fields.stringField(TABLE_NAME_FIELD, tableId.getName())));
+    });
   }
 
   public boolean isEmpty() {
-    return specMap.isEmpty();
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable registry = context.getTable(REGISTRY);
+      try (CloseableIterator<StructuredRow> it = registry.scan(Range.all(), Integer.MAX_VALUE)) {
+        return !it.hasNext();
+      }
+    });
+  }
+
+  private void createRegistryTable() throws IOException {
+    // Use a no-op registry to create PostgresSqlStructuredTableAdmin.
+    // During a table creation, except for registerSpecification, no other methods on the
+    // registry will be called by PostgresSqlStructuredTableAdmin.
+    StructuredTableRegistry noOpRegistry = new StructuredTableRegistry() {
+      final UnsupportedOperationException exception =
+        new UnsupportedOperationException("Not expected to be called during creation of registry!");
+
+      @Override
+      public void initialize() {
+        throw exception;
+      }
+
+      @Override
+      public void registerSpecification(StructuredTableSpecification specification) {
+        // Do nothing
+      }
+
+      @Nullable
+      @Override
+      public StructuredTableSpecification getSpecification(StructuredTableId tableId) {
+        throw exception;
+      }
+
+      @Override
+      public void removeSpecification(StructuredTableId tableId) {
+        throw exception;
+      }
+
+      @Override
+      public boolean isEmpty() {
+        throw exception;
+      }
+    };
+
+    try {
+      // Create the table if needed
+      PostgresSqlStructuredTableAdmin admin =
+        new PostgresSqlStructuredTableAdmin(noOpRegistry, dataSource);
+      if (!admin.tableExists(REGISTRY)) {
+        LOG.info("Creating SQL table {}", REGISTRY);
+        admin.create(SPEC);
+      }
+    } catch (TableAlreadyExistsException e) {
+      // Looks like the table was created concurrently by some other process
+      LOG.debug(String.format("Got exception when trying to create table %s", REGISTRY), e);
```

Comment:
Should we increase the log level of this error? I think we only do the initialization when we start CDAP right?

Reply:
I left it as debug since there is no action that needs to be taken when this happens